packages/SwingSet/src/kernel/kernel.js

```diff
@@ -786,108 +754,262 @@ export default function buildKernel(
+      assert(!terminationTrigger, 'hey, no delivery means no termination');
+    }
+
+    // terminate upon fatal syscalls, sys.exit requests, and worker problems
+    let didAbort = false;
+    if (terminationTrigger) {
+      assert(vatID, `terminationTrigger but not vatID`);
+      const ttvid = terminationTrigger.vatID;
+      assert.equal(ttvid, vatID, `wrong vat got terminated`);
+      const { abortCrank, reject, info } = terminationTrigger;
+      if (abortCrank) {
+        // errors unwind any changes the vat made
+        kernelKeeper.abortCrank();
+        didAbort = true;
+        // but metering deductions and underflow notifications must survive
+        if (deduction) {
+          const { meterID, computrons } = deduction; // re-deduct metering
+          const { notify } = kernelKeeper.deductMeter(meterID, computrons);
+          if (notify) {
+            notifyMeterThreshold(meterID); // re-queue notification
+          }
+        }
+        // some deliveries should be consumed when they fail
+        if (discardFailedDelivery) {
+          // kernelKeeper.abortCrank removed all evidence that the crank ever
+          // happened, including, notably, the removal of the delivery itself
+          // from the head of the run queue, which will result in it being
+          // delivered again on the next crank. If we don't want that, then
+          // we need to remove it again.
+
+          // eslint-disable-next-line no-use-before-define
+          getNextDeliveryMessage();
+        }
+        // other deliveries should be re-attempted on the next crank, so they
+        // get the right error: we leave those on the queue
+      }
+
+      // state changes reflecting the termination must also survive, so these
+      // happen after a possible abortCrank()
+      terminateVat(vatID, reject, info);
+      kernelSlog.terminateVat(vatID, reject, info);
+      kdebug(`vat terminated: ${JSON.stringify(info)}`);
+    }
+
+    if (!didAbort) {
+      // eslint-disable-next-line no-use-before-define
+      await vatWarehouse.maybeSaveSnapshot();
+    }
+    kernelKeeper.processRefcounts();
+    kernelKeeper.saveStats();
+    const crankNum = kernelKeeper.getCrankNumber();
+    kernelKeeper.incrementCrankNumber();
+    const { crankhash, activityhash } = kernelKeeper.commitCrank();
+    kernelSlog.write({
+      type: 'crank-finish',
+      crankNum,
+      crankhash,
+      activityhash,
+    });
+    return harden(policyInput);
+  }
+
+  let processQueueRunning;
+  async function tryProcessDeliveryMessage(message) {
+    if (processQueueRunning) {
+      console.error(`We're currently already running at`, processQueueRunning);
+      assert.fail(X`Kernel reentrancy is forbidden`);
+    }
+    processQueueRunning = Error('here');
+    try {
+      const policyInput = await processDeliveryMessage(message);
+      return policyInput;
     } finally {
       processQueueRunning = undefined;
     }
```

Comment:
Maybe add a note that the await is necessary for the `try` / `finally`. Not always obvious :)
Alternatively this is where a `.finally` comes in handy

```suggestion
    return processDeliveryMessage(message).finally(() => { processQueueRunning = undefined; });
```

Reply:
As long as that doesn't run afoul of our "don't trust possibly-doctored promises", I like it. In this case the promise is coming from our own code (and no user-provided Promises ever make it into the kernel), so we know it's safe.